### PR TITLE
Fix cppcheck 1.87 warnings in CurveFitting

### LINKED ISSUE #22912: Fix cppcheck 1.87 warnings
The newest version of cppcheck throws some warnings that need care.
[cppcheck1-84_report.txt](https://github.com/mantidproject/mantid/files/2191751/cppcheck1-84_report.txt)

### PR DESCRIPTION
**Pull request by soininen**
This fixes most `cppcheck` 1.87 issues in the CurveFitting module. A few cases where `cppcheck` suggested using an STL algorithm instead of a very simple range for loop it was deemed not worth it. These should be later suppressed by `// cppcheck-suppress useStlAlgorithm` when the build severs get an updated `cppcheck`.

**To test:**

Fixes #22912.

*This does not require release notes* because this is internal code maintenance only.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/CurveFitting/src/AugmentedLagrangianOptimizer.cpp [hunk 1]
```diff
@@ -222,7 +221,6 @@ void AugmentedLagrangianOptimizer::minimize(std::vector<double> &xv) const {
         break;
     }
     if (ICM == 0.0) {
-      ret = FTolReached;
```
Comment (DTasev) [inline Framework/CurveFitting/src/AugmentedLagrangianOptimizer.cpp]: Good spot, it's a large block of code! The behaviour seems unchanged, as `ret = Success;` [Old file, Line 210](https://github.com/mantidproject/mantid/pull/25213/files#diff-5a4677a945960304b4b1443a34c5cd27L210) was overwriting this in the next iteration anyway, and the `ret` value is only used in the `if` above.

File: Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp [hunk 2]
```diff
@@ -128,9 +128,6 @@ size_t CrystalFieldFunction::getNumberDomains() const {
   if (!m_target) {
     buildTargetFunction();
   }
-  if (!m_target) {
```
Comment (DTasev) [inline Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp]: Doesn't the removal of this check introduce a change for segfault on `m_target->getNumberDomains()`?

**Review by dtasev — CHANGES_REQUESTED** [review]
Just want to clarify one of the changes. Everything else looks good

File: Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp (hunk 2 above)
Comment (soininen) [reply Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp]: Good catch! I will revert this change.

**Review by dtasev — DISMISSED** [review]

File: Framework/CurveFitting/src/Functions/CrystalFieldMultiSpectrum.cpp [hunk 3]
```diff
@@ -125,9 +125,6 @@ size_t CrystalFieldMultiSpectrum::getNumberDomains() const {
   if (!m_target) {
     buildTargetFunction();
   }
-  if (!m_target) {
```
Comment (martyngigg) [inline Framework/CurveFitting/src/Functions/CrystalFieldMultiSpectrum.cpp]: I think we still need this check and exception for the same reason that @DTasev pointed out for `CrystalFieldFunction.cpp`?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Just one small fix that I think is needed?

**Review by martyngigg — APPROVED** [review]

**Pull request merged**